### PR TITLE
pgwire: migrate auth/conn logs to notable events

### LINKED ISSUE #56154: logging: enhance cockroach auth logging
With regards to Authorization errors, logs are split into two file currently:

- cockroach.sql-audit.log 
- cockroach-sql-exec.log. 

The `cockroach-sql-exec.log` has the actual authorization error, where as the audit log just has ERROR. 

We suggest to remove these event logs from above 2 files and create a new `cockroach.audit.log` that logs:

  - login attempts and successes, verbosely
  - enabled/disabled via a cluster setting, off by default

Linked to #56156

### PR DESCRIPTION
**Pull request by knz**
Fixes  #56154 

Release note (backward-incompatible change): The connection and
authentication logging enabled by the cluster settings
`server.auth_log.sql_connections.enabled` and
`server.auth_log.sql_sessions.enabled` was previously using a text
format which was hard to parse and integrate with external monitoring
tools. This has been changed to use the standard notable event
mechanism, with standardized payloads. The output format is now
structured; see its reference documentation for details about
the supported event types and payloads.

### CODE REVIEW

**Comment by knz** [comment]
cc @thtruo @logston

**Comment by knz** [comment]
friendly ping to the reviewers

**Review by otan — COMMENTED** [review]
> Reviewed 16 of 16 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57839)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @itsbilal, and @knz)

___
*[pkg/sql/pgwire/auth.go, line 358 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57839#-MQC1HIM18kVaWphpzLy:-MQC1HIM18kVaWphpzLz:b89es4v) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd19d532101b8f92724c357210ca605e8524ed7/pkg/sql/pgwire/auth.go#L358)):*
> ```Go
> 
> func (p *authPipe) AuthFail(err error) {
> 	p.readerDone <- authRes{err: err}
> ```

do we want to log here? seems easy to miss if AuthOK implictly logs but not AuthFail.
___
*[pkg/sql/pgwire/hba/hba.go, line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57839#-MQC1gXxEbAPtwRAf_cN:-MQC1gXxEbAPtwRAf_cO:bnjysc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd19d532101b8f92724c357210ca605e8524ed7/pkg/sql/pgwire/hba/hba.go#L94)):*
> ```Go
> 		return "host"
> 	default:
> 		panic(errors.Newf("unimplemented: %v", int(t)))
> ```

nit: "unimplemented conn type" (or something to identify better since i've been scared by panics omitting stack traces when "wrapped" by something else)
___
*[pkg/util/log/eventpb/session_events.proto, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57839#-MQC0lLK7CHkpeX2Bqq9:-MQC0lLK7CHkpeX2BqqA:b-84fsr2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd19d532101b8f92724c357210ca605e8524ed7/pkg/util/log/eventpb/session_events.proto#L57)):*
> ```protobuf
>   // The connection type after transport negotiation.
>   string transport = 1 [(gogoproto.jsontag) = ",omitempty"];
>   // The username the session is for.
> ```

should we specify whether this is normalised or not?
i notice we normalise this above -- is this something we want? (feels like we may lose information if we normalise them)

**Review by jordanlewis — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57839)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @itsbilal, and @knz)

___
*[pkg/sql/pgwire/auth.go, line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57839#-MQEi_rQ18n0-Oqj5Y_s:-MQEi_rQ18n0-Oqj5Y_t:b25xs69) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd19d532101b8f92724c357210ca605e8524ed7/pkg/sql/pgwire/auth.go#L114)):*
> ```Go
> 	tlsState, hbaEntry, methodFn, err := c.findAuthenticationMethod(authOpt)
> 	if err != nil {
> 		ac.LogAuthFailed(ctx, "authentication method lookup error", err)
> ```

This should be a warning and not an ordinary auth failed, right? or is this a user-specified auth method that might not exist?
___
*[pkg/sql/pgwire/auth.go, line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57839#-MQEijHJCStS0tXDbsqB:-MQEijHJCStS0tXDbsqC:b-o6vsmq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd19d532101b8f92724c357210ca605e8524ed7/pkg/sql/pgwire/auth.go#L126)):*
> ```Go
> 
> 	if err != nil {
> 		ac.LogAuthFailed(ctx, "authentication pre-hook failed", err)
> ```

what is this? should it be an warning instead? shouldn't AuthFailed be reserved for cases where users gave wrong credentials, and not just arbitrary errors?

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57839)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @itsbilal, @jordanlewis, and @otan)

___
*[pkg/sql/pgwire/auth.go, line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57839#-MQEi_rQ18n0-Oqj5Y_s:-MQMHq8W8x9_BMQtHXG-:b-895y71) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd19d532101b8f92724c357210ca605e8524ed7/pkg/sql/pgwire/auth.go#L114)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This should be a warning and not an ordinary auth failed, right? or is this a user-specified auth method that might not exist?
</blockquote></details>

To answer this point and the one below: we want an authn event logged in _all_ authn failure cases, regardless of cause. 

As to how this error can occur: This can naturally happen if none of the rules in the HBA configuration happen to match the incoming connection. That's a valid configuration (and pg logs a similar message in the same situation).
___
*[pkg/sql/pgwire/auth.go, line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57839#-MQEijHJCStS0tXDbsqB:-MQMIKm88li1oMV6duVS:b-l4gtbj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd19d532101b8f92724c357210ca605e8524ed7/pkg/sql/pgwire/auth.go#L126)):*
> what is this?

This happens when e.g. the method fails to recognize the parameters it needs in the connection, before the authn starts.

> should it be an warning instead? 

I'm not sure what you mean with “warning”? There's no notion of warning here: authentication either succeeds or fails, there's nothing in-between.

> shouldn't AuthFailed be reserved for cases where users gave wrong credentials, and not just arbitrary errors?

Ok now that's an interesting discussion: do we want more granular events *types* than authentication pass/fail with an "Info" field that describes the failure?
This requirement had not been communicated to me earlier. I was assuming we only needed authn ok/fail.
___
*[pkg/sql/pgwire/auth.go, line 358 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57839#-MQC1HIM18kVaWphpzLy:-MQMK0yY2Sxj23a4ckS7:b-23u76i) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd19d532101b8f92724c357210ca605e8524ed7/pkg/sql/pgwire/auth.go#L358)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

do we want to log here? seems easy to miss if AuthOK implictly logs but not AuthFail.
</blockquote></details>

Yeah that was not a great idea. Moved the logging for the OK case side-by-side with the fail case instead, to improve legibility.
___
*[pkg/sql/pgwire/hba/hba.go, line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57839#-MQC1gXxEbAPtwRAf_cN:-MQMKAHB9jMfS1K5tpbY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd19d532101b8f92724c357210ca605e8524ed7/pkg/sql/pgwire/hba/hba.go#L94)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nit: "unimplemented conn type" (or something to identify better since i've been scared by panics omitting stack traces when "wrapped" by something else)
</blockquote></details>

Done.
___
*[pkg/util/log/eventpb/session_events.proto, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57839#-MQC0lLK7CHkpeX2Bqq9:-MQMKULO8jqA-5xR3DMe:b-rkcegm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd19d532101b8f92724c357210ca605e8524ed7/pkg/util/log/eventpb/session_events.proto#L57)):*
> should we specify whether this is normalised or not?

Yes sure, done.

>  i notice we normalise this above -- is this something we want? (feels like we may lose information if we normalise them)

the normalization occurs in fact very early during the conn handshake. the `.Normalized()` call merely returns the stored normalized username. By this time it's too late to log the original.

(I'll ask Aaron if logging the original usernam eis important)

**Review by otan — APPROVED** [review]
> Reviewed 7 of 8 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57839)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @itsbilal, @jordanlewis, and @otan)

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57839)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl, @arulajmani, @itsbilal, @jordanlewis, and @otan)

___
*[pkg/sql/pgwire/auth.go, line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57839#-MQEijHJCStS0tXDbsqB:-MQO6_0t4bj9g6pcCEgv:be6wfm7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd19d532101b8f92724c357210ca605e8524ed7/pkg/sql/pgwire/auth.go#L126)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

> what is this?
This happens when e.g. the method fails to recognize the parameters it needs in the connection, before the authn starts.

> should it be an warning instead? 
I'm not sure what you mean with “warning”? There's no notion of warning here: authentication either succeeds or fails, there's nothing in-between.

> shouldn't AuthFailed be reserved for cases where users gave wrong credentials, and not just arbitrary errors?
Ok now that's an interesting discussion: do we want more granular events *types* than authentication pass/fail with an "Info" field that describes the failure?
This requirement had not been communicated to me earlier. I was assuming we only needed authn ok/fail.
</blockquote></details>

Discussed with @aaron-crl : the Reason field is now populated with an enum with documented + stable values, instead of an unscrupulous string.
___
*[pkg/util/log/eventpb/session_events.proto, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57839#-MQC0lLK7CHkpeX2Bqq9:-MQO6hJwCx4iVU9uGth-:b-t8zhdh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fd19d532101b8f92724c357210ca605e8524ed7/pkg/util/log/eventpb/session_events.proto#L57)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

> should we specify whether this is normalised or not?

Yes sure, done.

>  i notice we normalise this above -- is this something we want? (feels like we may lose information if we normalise them)

the normalization occurs in fact very early during the conn handshake. the `.Normalized()` call merely returns the stored normalized username. By this time it's too late to log the original.

(I'll ask Aaron if logging the original usernam eis important)
</blockquote></details>

Discussed with the team - this can be added in a separate PR.

**Comment by knz** [comment]
TFYRs!

bors r=otan,jordanlewis
